### PR TITLE
Added colored output, json output. And removed unnecessary log prefixes.

### PR DESCRIPTION
**Pull request by AnikHasibul**
Added log message color with `github.com/logrusorgru/aurora` and prettified the log output!

### CODE REVIEW

File: cmd/gocritic/main.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func main() {
 	if cmd == nil {
 		terminate("unknown sub-command: "+sub, printSupportedSubs)
 	}
-	log.SetPrefix(sub + ": ")
+	log.SetPrefix("- ")
```
Comment (Quasilyte) [inline cmd/gocritic/main.go]: Maybe remove prefix completely then?

File: cmd/criticize/main.go [hunk 2]
```diff
@@ -285,7 +286,13 @@ func (l *linter) checkFile(f *ast.File) {
...
 					loc = shortenLocation(loc)
 				}
-				log.Printf("%s: %s: %v\n", loc, c.Rule, warn.Text)
+				log.Printf("%v:\n%v: %v\n",
+					aurora.Red(c.Rule),
+					aurora.Magenta(
+						aurora.Bold(loc),
+					),
+					warn.Text,
+				)
```
Comment (Quasilyte) [inline cmd/criticize/main.go]: Colored output should be optional and controlled by a flag.
It's also undesirable to break 1-line per warning rule (although some warnings are printing multi-line code, but this is a bug).
It also changes `location: checker: text` to `checker: location: text`. Not sure whether it's better.

File: cmd/gocritic/main.go (hunk 1 above)
Comment (AnikHasibul) [reply cmd/gocritic/main.go]: Just putting a `-` to make it more understandable that "Here is a new log!"

File: cmd/criticize/main.go (hunk 2 above)
Comment (AnikHasibul) [reply cmd/criticize/main.go]:
```txt
checker:
loc/a/tion: text
```

```txt
- ifElseChain:
/home/go/src/logger/text_formatter.go:129:2: should rewrite if-else to switch statement
```

Doesn't this formation look better?

Flag for non-colored output would be great. But also thinking about a `json` flag to work with various editor. As always `vim-go` is the first choice.

File: cmd/gocritic/main.go (hunk 1 above)
Comment (cristaloleg) [reply cmd/gocritic/main.go]: Anyway we're expecting a log, so `-` doesn't provides useful context.

File: cmd/criticize/main.go (hunk 2 above)
Comment (cristaloleg) [reply cmd/criticize/main.go]: I think it should be configurable. Having oneline output simplifies work with a log, not everyone want to see 2 or more lines per warning.
Consider to make it configurable 😉 

> But also thinking about a json flag to work with various editor.

That's a good one, probably. Do you know which other linters provides a json output?

> As always vim-go is the first choice.

Huh, debatable ;D

File: cmd/criticize/main.go [hunk 3]
```diff
@@ -231,14 +233,22 @@ func (l *linter) CheckPackage(pkgPath string) {
 
 	l.ctx.SetPackageInfo(&pkgInfo.Info, pkgInfo.Pkg)
 	for _, f := range pkgInfo.Files {
-		if l.flags.CheckGenerated || !isGenerated(f) {
-			l.ctx.SetFileInfo(l.getFilename(f))
-			l.checkFile(f)
+		if l.flags.IgnoreTests && l.isTestFile(f) {
```
Comment (cristaloleg) [inline cmd/criticize/main.go]: This is strange. Those changes were already merged into master 🤔

File: cmd/criticize/main.go [hunk 4]
```diff
@@ -285,7 +295,41 @@ func (l *linter) checkFile(f *ast.File) {
...
+					)
+
+				case l.flags.JSONOutput:
+
+					// due to sort capabilities,
+					// using struct instead of map
+					b, err := json.MarshalIndent(struct {
+						Rule     string
+						Location string
+						Warn     string
```
Comment (cristaloleg) [inline cmd/criticize/main.go]: JSON tags?

File: cmd/criticize/main.go [hunk 5]
```diff
@@ -285,7 +295,41 @@ func (l *linter) checkFile(f *ast.File) {
 				if l.flags.ShorterErrLocation {
 					loc = shortenLocation(loc)
 				}
-				log.Printf("%s: %s: %v\n", loc, c.Rule, warn.Text)
+				switch {
```
Comment (cristaloleg) [inline cmd/criticize/main.go]: What about extracting this `switch` into a func? WDYT?

File: cmd/criticize/main.go [hunk 6]
```diff
@@ -285,7 +295,41 @@ func (l *linter) checkFile(f *ast.File) {
...
+
+					// due to sort capabilities,
+					// using struct instead of map
+					b, err := json.MarshalIndent(struct {
+						Rule     string
+						Location string
+						Warn     string
+					}{
+						// Using value to ignore c.Rule.Docs and other unnecessary fields
+						Rule:     fmt.Sprintf("%v", c.Rule),
```
Comment (cristaloleg) [inline cmd/criticize/main.go]: `c.Rule.Name()` will be enough 😉

File: cmd/gocritic/main.go [hunk 7]
```diff
@@ -62,7 +62,7 @@ func main() {
 	if cmd == nil {
 		terminate("unknown sub-command: "+sub, printSupportedSubs)
 	}
-	log.SetPrefix(sub + ": ")
+	log.SetPrefix("")
```
Comment (cristaloleg) [inline cmd/gocritic/main.go]: I think default prefix is an empty string already, isn't it?

File: cmd/criticize/main.go [hunk 8]
```diff
@@ -302,3 +311,41 @@ func shortenLocation(loc string) string {
...
+
+func output(l *linter, rule, loc, warn string) {
+	switch {
+	case l.flags.JSONOutput:
+		// due to sort capabilities,
+		// using struct instead of map
+		b, err := json.MarshalIndent(struct {
+			Rule     string `json:"rule"`
+			Location string `json:"location"`
+			Warning  string `jdon:"warning"`
```
Comment (Quasilyte) [inline cmd/criticize/main.go]: `s/jdon/json/`

File: cmd/criticize/main.go [hunk 9]
```diff
@@ -302,3 +311,41 @@ func shortenLocation(loc string) string {
...
+			),
+			warn,
+		)
+
+	default:
+		log.Printf("%s:\n%s: %s\n",
+			rule,
+			loc,
+			warn,
+		)
```
Comment (Quasilyte) [inline cmd/criticize/main.go]: Adding new formats might be okay, but why changing the default one?
There is no "right default format", and I personally don't like a newline inside a warning message.

File: cmd/gocritic/main.go (hunk 7 above)
Comment (Quasilyte) [reply cmd/gocritic/main.go]: I think the call to `log.SetPrefix` should be removed completely.

**Comment by Quasilyte** [comment]
@AnikHasibul, sorry for long delay. I've been quite busy these days, unfortunately (but this is a no excuse for me).

I want to share my thoughts:
1. It's somewhat distressing that this PR seem to have broken merge with master, as diff shows previously merged code. Could we fix this? If it's unclear how to do it in this PR, maybe pull locally and do a new one? Would be far easier and comfortable.
2. Can we do these 3 things in a separate PRs?
2.1 Removing log prefix. This can be merged immediately. No objections.
2.2 Custom message formatting (or JSON+plain text). I have some ideas, will share with them in form of an issue, where we can discuss it.
2.3 Message colors.

Thanks.

**Comment by AnikHasibul** [comment]
Thanks for the clearance!
Closing this PR. I am opening a new one!

**Comment by AnikHasibul** [comment]
New PR #622

**Pull request closed without merging**